File: djJava/HelloJava06/src/javastudy02_abstract/Animal.java
package javastudy02_abstract;

//추상 클래스 : 추상적인 클래스, 추상적인 개념
//직접적으로 인스턴스 생성이 불가능
//Animal a = new Animal(); // <- 이거 안 됨

//'동물'이라는 이름의 동물은 없다
//다만 '동물' 특징을 갖는 개, 고양이 등등이 있음...
public abstract class Animal {
	
	private String name;
	private int age;
	public static String color; //클래스변수나 클래스 메서드는 
	//조상클래스와 자손 클래스가 모두 공유를 함
	public String getName() {
		return name;
	}
	public void setName(String name) {
		this.name = name;
	}
	public int getAge() {
		return age;
	}
	public void setAge(int age) {
		this.age = age;
	}
	public abstract void sleep(); //Animal을 상속받은 클래스에서
 	public abstract void eat(); //반드시 구현해야 하는 것들!!
	public void born() {
		System.out.println("응애 응애 태어났어요!");
	}

}

File: djCSharp2/HelloCSharp004/HelloCSharp004/Animal.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace HelloCSharp004
{
    //abstract 클래스 = 추상 클래스
    //자신 가지고 직접 인스턴스 못 만듦
    //즉 Animal a = new Animal() 이런 거 못 함
    //대신 Animal을 상속받은 클래스에서는 메서드 구체화 가능하고
    //인스턴스도 만들 수 있음
    public abstract class Animal
    {
        public string name { get; set; }
        public int age { get; set; }
        public void sleep() //기본으로 구현된 기능
        {
            Console.WriteLine(name+" : 잠을 잡니다.");
        }
        public abstract void eat(); //abstract가 붙은 건
        //상속받은 클래스마다 모두 각자 다르게 eat을 구현해야 하고
        //또한 eat을 무조건 구현해야 함!!!
        public abstract void fight();
        public virtual void run() //Animal을 상속받았다면 오버라이드 가능
        {
            Console.WriteLine(age+"살 답게 도망갑니다.");
        }
    }
}
